src/engine/Server.cpp

```diff
@@ -163,47 +164,68 @@ void Server::run(const string& indexBaseName, bool useText, bool usePatterns,
+    }
+    return parsedUrl;
   }
   if (request.method() == http::verb::post) {
     // For a POST request, the content type *must* be either
-    // "application/x-www-form-urlencoded" or "application/sparql-query". In
-    // the first case, the body of the POST request contains a URL-encoded
-    // query (just like in the part of a GET request after the "?"). In the
-    // second case, the body of the POST request contains *only* the SPARQL
-    // query, but not URL-encoded, and no other URL parameters. See Sections
-    // 2.1.2 and 2.1.3 of the SPARQL 1.1 standard:
-    // https://www.w3.org/TR/2013/REC-sparql11-protocol-20130321
+    // "application/x-www-form-urlencoded" (1) or "application/sparql-query"
+    // (2).
+    //
+    // (1) Section 2.1.2: The body of the POST request contains a URL-encoded
+    // query (just like in the part of a GET request after the "?").
+    //
+    // (2) Section 2.1.3: The body of the POST request contains *only* the
+    // SPARQL query, but not URL-encoded, and no other URL parameters.
+    //
+    // Reference: https://www.w3.org/TR/2013/REC-sparql11-protocol-20130321
     std::string_view contentType = request.base()[http::field::content_type];
     LOG(DEBUG) << "Content-type: \"" << contentType << "\"" << std::endl;
     static constexpr std::string_view contentTypeUrlEncoded =
         "application/x-www-form-urlencoded";
     static constexpr std::string_view contentTypeSparqlQuery =
         "application/sparql-query";
 
-    // In either of the two cases explained above, we convert the data to a
-    // format as if it came from a GET request. The second argument to
-    // `parseGetRequestTarget` says whether the function should apply URL
-    // decoding.
     // Note: For simplicity we only check via `starts_with`. This ignores
     // additional parameters like `application/sparql-query;charset=utf8`. We
     // currently always expect UTF-8.
     // TODO<joka921> Implement more complete parsing that allows the checking of
     // these parameters.
     if (contentType.starts_with(contentTypeUrlEncoded)) {
-      return ad_utility::UrlParser::parseGetRequestTarget(
-          absl::StrCat(toStd(request.target()), "?", request.body()), true);
+      // Query String Parameters must be empty for URL-encoded POST. See
+      // SPARQL 1.1 Protocol Sections 2.1 and 3.1
+      if (!parsedUrl.parameters_.empty()) {
+        throw std::runtime_error(
+            "URL-encoded POST requests must not contain query parameters in "
+            "the URL.");
+      }
+      // Set the parameters from the request body.
+      parsedUrl.parameters_ = ad_utility::UrlParser::paramsToMap(
+          boost::urls::parse_query(request.body()).value());
+
+      if (parsedUrl.parameters_.contains("query")) {
+        parsedUrl.query_ = parsedUrl.parameters_["query"];
+      } else {
+        throw std::runtime_error("POST Request missing \"query\" parameter.");
+      }
+
+      return parsedUrl;
```

Comment:
I think The interface of the `nullopt` in the `parsedUrl` is currently overloaded. 
It currently can either mean "no query" (e.g. because it only was a `clear-cache` request) or `the query was not yet set, as we are still parsing`. 
This can be disambiguated by having two types: a `ParsedUrl`(or whatever the name is) with simply a target and parameters, and a `ParsedUrlWithQuery`, that additionally has the (still optional) query. This function here then converts the one into the other, and everyone is happy.